Some fixes to address issue #9 Restricted Admin Mode & added check if the current IdentityReference is actually a SID

techspence opened the pull request:
This pr consists of the following changes to resolve https://github.com/TrimarcJake/Locksmith/issues/9

- A new function `Get-RestrictedAdminModeSetting` with the purpose of checking the registry to see if Restricted Admin Mode (RAM) is enabled
  - The function checks for RAM in the following way:
	- if `HKLM:SYSTEM\CurrentControlSet\Control\Lsa\DisableRestrictedAdminMode` is set to 0 and `HKLM:SYSTEM\CurrentControlSet\Control\Lsa\DisableRestrictedAdminOutboundCreds` is set to 1 then RAM is enabled

- If RAM is found to be enabled, the script now warns you, tells you to use `-Credential` (a new argument added in this pr) and exits
  - `-Credential` is needed when RAM is enabled so credentials can be captured and passed to Get-AD cmdlets such as `Get-ADForest`, etc.

- A new argument has been added: `-Credential`
  - The intended usage is to add the following to the execution of Locksmith: `-Credential domain\username`

- Along with the new argument to the script as a whole, the -Credential option has been added to the following functions:
  - `Get-Target`
  - `Get-ADCSObject`
  - `Set-AdditionalCAProperty`
  - `Get-CAHostObject`

- When passing the `-Credential` option, `certutil.exe` no longer works. To remedy that, Locksmith has been updated to check the AuditFilter and EditFlags via the registry on the CA server using Invoke-Command

- Small fix related to https://github.com/TrimarcJake/Locksmith/issues/15 to check if the current `IdentityReference` is actually a SID and if it is, don't translate it to a SID, well because it's already a SID. :)

I think that's it. Please test and let me know if I broke anything or if anything doesn't work :)

TrimarcJake commented:
My testing reveals the following:
1. The Credential flag works great for the spots where it's implemented! 
2. The warning about certutil fixes not working in Restricted Mode should not appear in Mode 0 at all since no fixes are presented.
3. The certutil fixes warning should only appear when there are Auditing or ESC6 issues.
4. The certutil fixes warning should appear directly beneath the relevant section(s) instead of at the very end of the output.
5. The registry check for the SAN flag doesn't seem to work. I'm always seeing that the flag is enabled when it is definitely not. The correct registry value is being captured, but the logic for setting $SANFlag is wrong (lines 201-209).
6. Related to 5, the registry check for the SAN flag is not robust enough because it's only checking for one specific configuration of all the flags. It should instead be a bitwise comparison. 

If you don't want to figure out the bitwise math, I think we can safely revert to using certutil.exe by wrapping the existing commands in Invoke-Command and passing the $Credential to Invoke-Command. Something like this
``` powershell
Invoke-Command -Credential $Credential -ComputerName $CAHostName -ScriptBlock { certutil -config $CAFullName -getreg policy\EditFlags }
```
Wrapping certutil with Invoke-Command should work for the auditing & ESC6 fixes as well. :D 

Also, since you did so much work on this, I'm willing to handle these fixes.

techspence commented:
This ended up being quicker to resolve than I expected. Fixes pushed in https://github.com/TrimarcJake/Locksmith/pull/16/commits/02152601ef6b1a95c303cd3d8ec0ac4b9994a197 Please let me know if any of these changes break.

- Fixed minor logic issue in Get-RestrictedAdminModeSetting

2. **Resolved** ~~The warning about certutil fixes not working in Restricted Mode should not appear in Mode 0 at all since no fixes are presented.~~

3. **Resolved** ~~The certutil fixes warning should only appear when there are Auditing or ESC6 issues.~~

4. **Resolved, feedback requested** ~~The certutil fixes warning should appear directly beneath the relevant section(s) instead of at the very end of the output.~~
	- Resolved, but i'm unsure I like how this looks, aesthetically. This could be moved to directly underneath the name of the section and it may look better. Thoughts?

5. **Resolved** ~~The registry check for the SAN flag doesn't seem to work. I'm always seeing that the flag is enabled when it is definitely not. The correct registry value is being captured, but the logic for setting $SANFlag is wrong (lines 201-209).~~

6. **Resolved** ~~Related to 5, the registry check for the SAN flag is not robust enough because it's only checking for one specific configuration of all the flags. It should instead be a bitwise comparison.~~
	- Resolved 5 and 6 by wrapping certutil in invoke-command when restricted admin mode is detected
	- I reverted to the original logic for the AuditFilter and SANFlag checks
